narwhals/expr.py

```diff
@@ -3605,6 +3606,96 @@ def sum_horizontal(*exprs: IntoExpr | Iterable[IntoExpr]) -> Expr:
     )
 
 
+class When:
+    def __init__(self, condition: Expr) -> None:
+        self._condition = condition
+
+    def then(self, value: Any) -> Then:
+        return Then(lambda plx: plx.when(self._condition._call(plx)).then(value))
+
+
+class Then(Expr):
+    def __init__(self, call: Callable[[Any], Any]) -> None:
+        self._call = call
+
+    def otherwise(self, value: Any) -> Expr:
+        return Expr(lambda plx: self._call(plx).otherwise(value))
+
+    def when(
+        *predicates: IntoExpr | Iterable[IntoExpr],
+        **constraints: Any,  # noqa: ARG002
+    ) -> ChainedWhen:
+        return ChainedWhen(reduce(lambda a, b: a & b, flatten([predicates])))
+
+
+class ChainedWhen:
+    def __init__(self, condition: Expr) -> None:
+        self._condition = condition
+
+    def then(self, value: Any) -> ChainedThen:
+        return ChainedThen(lambda plx: plx.when(self._condition._call(plx)).then(value))
+
+
+class ChainedThen(Expr):
+    def __init__(self, call: Callable[[Any], Any]) -> None:
+        self._call = call
+
+    def otherwise(self, value: Any) -> Expr:
+        return Expr(lambda plx: self._call(plx).otherwise(value))
+
+
+def when(*predicates: IntoExpr | Iterable[IntoExpr], **constraints: Any) -> When:  # noqa: ARG001
+    """
+    Start a `when-then-otherwise` expression.
+    Expression similar to an `if-else` statement in Python. Always initiated by a `pl.when(<condition>).then(<value if condition>)`., and optionally followed by chaining one or more `.when(<condition>).then(<value>)` statements.
+    Chained when-then operations should be read as Python `if, elif, ... elif` blocks, not as `if, if, ... if`, i.e. the first condition that evaluates to `True` will be picked.
+    If none of the conditions are `True`, an optional `.otherwise(<value if all statements are false>)` can be appended at the end. If not appended, and none of the conditions are `True`, `None` will be returned.
+
+    Parameters:
+        predicates
+            Condition(s) that must be met in order to apply the subsequent statement. Accepts one or more boolean expressions, which are implicitly combined with `&`. String input is parsed as a column name.
+        constraints
+            Apply conditions as `col_name = value` keyword arguments that are treated as equality matches, such as `x = 123`. As with the predicates parameter, multiple conditions are implicitly combined using `&`.
+
+    Examples:
+        >>> import pandas as pd
+        >>> import polars as pl
+        >>> import narwhals as nw
+        >>> df_pl = pl.DataFrame({"a": [1, 2, 3], "b": [5, 10, 15]})
+        >>> df_pd = pd.DataFrame({"a": [1, 2, 3], "b": [5, 10, 15]})
+
+        We define a dataframe-agnostic function:
+
+        >>> @nw.narwhalify
+        ... def func(df_any):
+        ...     from narwhals.expr import when
+        ...
+        ...     return df_any.with_columns(
+        ...         when(nw.col("a") < 3).then(5).otherwise(6).alias("a_when")
```

Comment:
I think we should export it from the top-level, so then here it becomes `nw.when` (just like in Polars you'd do `pl.when`)

Reply:
I have added it to the top level api, however is not yet stabalised, so there are some errors in the test suit